File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        // row
        for(int i=0; i<9; i++){
            set<char> s;
            for(int j=0; j<9; j++){
                if(s.find(board[i][j]) != s.end()) return false;
                if(board[i][j] != '.') s.insert(board[i][j]);
            }
        }
        // column
        for(int i=0; i<9; i++){
            set<char> s;
            for(int j=0; j<9; j++){
                if(s.find(board[j][i]) != s.end()) return false; 
                if(board[j][i] != '.') s.insert(board[j][i]);
            }
        }
        // box

        vector<pair<int, int>> start = {
            {0, 0}, {0, 3}, {0, 6},
            {3, 0}, {3, 3}, {3, 6},
            {6, 0}, {6, 3}, {6, 6}
        };
        for (const auto& p : start) {
            int i = p.first;
            int j = p.second;
            set<char> s;
            for (int row = i; row < i + 3; row++) {
                for (int col = j; col < j + 3; col++) {
                    if(s.find(board[row][col]) != s.end()) return false;
                    if(board[row][col] != '.') s.insert(board[row][col]);
                }   
            }
        }
    return true;
    }
};

File: 0036-valid-sudoku/0036-valid-sudoku.kt
class Solution {
    fun isValidSudoku(board: Array<CharArray>): Boolean {
        // Row check
        for (i in 0 until 9) {
            val seen = mutableSetOf<Char>()
            for (j in 0 until 9) {
                val c = board[i][j]
                if (c != '.') {
                    if (c in seen) return false
                    seen.add(c)
                }
            }
        }

        // Column check
        for (i in 0 until 9) {
            val seen = mutableSetOf<Char>()
            for (j in 0 until 9) {
                val c = board[j][i]
                if (c != '.') {
                    if (c in seen) return false
                    seen.add(c)
                }
            }
        }

        // Box check
        val starts = listOf(
            Pair(0, 0), Pair(0, 3), Pair(0, 6),
            Pair(3, 0), Pair(3, 3), Pair(3, 6),
            Pair(6, 0), Pair(6, 3), Pair(6, 6)
        )

        for ((i, j) in starts) {
            val seen = mutableSetOf<Char>()
            for (row in i until i + 3) {
                for (col in j until j + 3) {
                    val c = board[row][col]
                    if (c != '.') {
                        if (c in seen) return false
                        seen.add(c)
                    }
                }
            }
        }

        return true
    }
}
